### PR TITLE
SQ-55/notification service

### PR DESCRIPTION
**Pull request by danybony**
### Description
This adds an intent service responsible of showing the about-to-start notification for event starting in the next 10 minutes.

### Details
First of all I added `start` and `end` Date fields to `Event`.
I added the IntentService responsible of getting all the favourite events (this will be done after we implement the favourite events logic) and displaying the notification(s) for all the events starting in the next 10 minutes.
Also, the service re-schedule itself for 10 minutes before the next event starting at least in 10 minutes from now (the first not included in the previous batch of notifications).
The IntentService is first started with a BOOT_COMPLETED broadcast receiver.

**Bonus**: added the `Chronometer` style to notifications, which automatically adds a countdown for the event start.

#### Still TODO
* Create the events with the right start and end
* Implement the favourite events logic

### Screenshot
You can see the elapsed time from the event start next to the app name in the notification.
(It will be negative in the minutes before the event start)
device-2017-02-24-174841

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -58,6 +62,8 @@
       </intent-filter>
     </service>
 
+    <service android:name=".notification.NotificationService" />
```
Comment (rock3r) [inline app/src/main/AndroidManifest.xml]: Would be nice to have [direct boot](https://developer.android.com/training/articles/direct-boot.html) support too. Maybe you could also explicitly set this to `exported=false` so it's clearer?

File: app/src/main/java/net/squanchy/notification/NotificationService.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    private static final int NOTIFICATION_INTERVAL_MINUTES = 10;
+
+    public NotificationService() {
+        super(NotificationService.class.getSimpleName());
+    }
+
+    @Override
+    protected void onHandleIntent(Intent intent) {
+        NotificationCreator notificationCreator = new NotificationCreator(this);
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/notification/NotificationService.java]: This and the notifier could be initialised in the ctor, no?

File: app/src/main/java/net/squanchy/notification/NotificationService.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+        sortedFavourites
+                .filter(event -> event.start().after(notificationIntervalEnd))
+                .take(1)
+                .subscribe(this::scheduleNextAlarm);
+    }
+
+    private boolean startingIn(Date now, Date notificationIntervalEnd, Event event) {
+        Date eventStart = event.start();
+        return now.before(eventStart) && (eventStart.before(notificationIntervalEnd) || eventStart == notificationIntervalEnd);
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/notification/NotificationService.java]: Maybe extract the two sides of the `&&` so it's more readable?

PS: `eventStart.after(now)` is more readable than `now.before(eventStart)`

File: app/src/main/java/net/squanchy/notification/NotificationService.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                .map(notificationCreator::createFrom)
+                .subscribe(notifier::showNotifications);
+
+        sortedFavourites
+                .filter(event -> event.start().after(notificationIntervalEnd))
+                .take(1)
+                .subscribe(this::scheduleNextAlarm);
+    }
+
+    private boolean startingIn(Date now, Date notificationIntervalEnd, Event event) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/notification/NotificationService.java]: There's three arguments, the "in" suffix is confusing. Maybe `startingBeforeEndOfInterval`?
You could even split out into two `filters`, one for the "starts in the future" and one for "starts within interval".

File: app/src/main/java/net/squanchy/schedule/domain/view/Event.java [hunk 5]
```diff
@@ -16,6 +17,8 @@
     public static Event create(
             long eventId,
             int dayId,
+            Date start,
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/schedule/domain/view/Event.java]: ❌ Joda time use you must

**Review by rock3r — CHANGES_REQUESTED** [review]

File: app/src/main/java/net/squanchy/notification/NotificationService.java (hunk 4 above)
Comment (danybony) [reply app/src/main/java/net/squanchy/notification/NotificationService.java]: Yes, splitting the two filters would make it more readable

File: app/src/debug/java/net.squanchy/DebugActivity.java [hunk 6]
```diff
@@ -63,9 +64,9 @@ private void createAndNotifyTalksCount(int count) {
     private Event createTestEvent(int id) {
         Calendar calendar = Calendar.getInstance();
         calendar.add(Calendar.MINUTE, 5);
-        Date start = calendar.getTime();
+        DateTime start = new DateTime(calendar.getTime());
```
Comment (rock3r) [inline app/src/debug/java/net.squanchy/DebugActivity.java]: ❌ don't use Calendar (Java's own `Date` APIs) mixed with JodaTime's `DateTime`.

Here you should have:

```java
DateTime start = new DateTime().plusMinutes(5);
DateTime end = start.plusMinutes(45);
```

File: app/src/main/AndroidManifest.xml [hunk 7]
```diff
@@ -62,7 +63,9 @@
       </intent-filter>
     </service>
 
-    <service android:name=".notification.NotificationService" />
+    <service
+      android:name=".notification.NotificationService"
```
Comment (rock3r) [inline app/src/main/AndroidManifest.xml]: Missed before too, but use fully qualified name plz

File: app/src/main/java/net/squanchy/notification/NotificationService.java [hunk 8]
```diff
@@ -6,64 +6,58 @@
...
     @Override
     protected void onHandleIntent(Intent intent) {
-        NotificationCreator notificationCreator = new NotificationCreator(this);
-        Notifier notifier = Notifier.from(this);
-
-        Calendar calendar = Calendar.getInstance();
-        calendar.add(Calendar.MINUTE, NOTIFICATION_INTERVAL_MINUTES);
-        Date notificationIntervalEnd = calendar.getTime();
-        Date now = new Date();
+        DateTime notificationIntervalEnd = new DateTime().withPeriodAdded(NOTIFICATION_INTERVAL, 1);
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/notification/NotificationService.java]: Why using the Period thing with a count of one and not `plus(NOTIFICATION_INTERVAL)`?

PS: you probably don't want to use a `ReadablePeriod` such as `Minutes` when adding, according to the `plus()` docs:

>This method is typically used to add complex period instances. Adding one field is best achieved using methods like `#plusYears(int)`.

You can use `plusMinutes()` and leave the constant being an `int`

File: app/src/main/java/net/squanchy/notification/NotificationService.java [hunk 9]
```diff
@@ -6,64 +6,58 @@
...
 
         Observable<Event> favourites = Observable.empty(); // TODO load all events from somewhere
         Observable<Event> sortedFavourites = favourites.sorted(byStartDate());
 
         sortedFavourites
-                .filter(event -> startingIn(now, notificationIntervalEnd, event))
+                .filter(event -> event.start().isAfterNow())
+                .filter(event -> event.start().isBefore(notificationIntervalEnd) || event.start().isEqual(notificationIntervalEnd))
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/notification/NotificationService.java]: Extract `isBeforeOrEqualTo()` method?

File: app/src/main/java/net/squanchy/notification/NotificationService.java [hunk 10]
```diff
@@ -6,64 +6,58 @@
...
-
     private Comparator<Event> byStartDate() {
         return (event1, event2) -> event1.start().compareTo(event2.start());
     }
 
     private void scheduleNextAlarm(Event event) {
-        Calendar calendar = Calendar.getInstance();
-        calendar.setTime(event.start());
-        calendar.add(Calendar.MINUTE, -NOTIFICATION_INTERVAL_MINUTES);
+        DateTime serviceAlarm = event.start().withPeriodAdded(NOTIFICATION_INTERVAL, -1);
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/notification/NotificationService.java]: Same goes here, use `minus()` instead

**Review by rock3r — CHANGES_REQUESTED** [review]

File: app/src/main/java/net/squanchy/notification/NotificationService.java (hunk 8 above)
Comment (danybony) [reply app/src/main/java/net/squanchy/notification/NotificationService.java]: Ah, good to know, I totally missed that `plusMinutes()` method

File: app/src/debug/java/net.squanchy/DebugActivity.java [hunk 11]
```diff
@@ -62,11 +61,8 @@ private void createAndNotifyTalksCount(int count) {
     }
 
     private Event createTestEvent(int id) {
-        Calendar calendar = Calendar.getInstance();
-        calendar.add(Calendar.MINUTE, 5);
-        DateTime start = new DateTime(calendar.getTime());
-        calendar.add(Calendar.MINUTE, 45);
-        DateTime end = new DateTime(calendar.getTime());
+        DateTime start = new DateTime().plusMinutes(5);
+        DateTime end = new DateTime().plusMinutes(45);
```
Comment (rock3r) [inline app/src/debug/java/net.squanchy/DebugActivity.java]: This is _technically_ different from what you had before (the new date will be a few millis after the first one) but hey it's test code XD

**Review by rock3r — APPROVED** [review]
Shipit 🚢

**Pull request merged**